src/coprocessor/codec/mysql/duration.rs

```diff
@@ -13,215 +13,316 @@
 
 use crate::util::codec::number::{self, NumberEncoder};
 use crate::util::codec::BytesSlice;
+use bitfield::bitfield;
 use std::cmp::Ordering;
 use std::fmt::{self, Display, Formatter};
 use std::io::Write;
 use std::time::Duration as StdDuration;
 use std::{i64, str, u64};
-use time::{self, Tm};
 
 use super::super::Result;
 use super::{check_fsp, parse_frac, Decimal};
 
-pub const NANOS_PER_SEC: i64 = 1_000_000_000;
+pub const NANOS_PER_SEC: u64 = 1_000_000_000;
 pub const NANO_WIDTH: u32 = 9;
 const SECS_PER_HOUR: u64 = 3600;
 const SECS_PER_MINUTE: u64 = 60;
+const MINUTE_PER_HOUR: u64 = 60;
+const NANOS_PER_MICRO: u64 = 1_000;
 
 /// `MAX_TIME_IN_SECS` is the maximum for mysql time type.
 const MAX_TIME_IN_SECS: u64 = 838 * SECS_PER_HOUR + 59 * SECS_PER_MINUTE + 59;
 
-fn check_dur(dur: &StdDuration) -> Result<()> {
-    let secs = dur.as_secs();
-    if secs > MAX_TIME_IN_SECS || secs == MAX_TIME_IN_SECS && dur.subsec_nanos() > 0 {
+fn check_dur(duration: &StdDuration) -> Result<()> {
+    let secs = duration.as_secs();
+    if secs > MAX_TIME_IN_SECS || secs == MAX_TIME_IN_SECS && duration.subsec_nanos() > 0 {
         return Err(invalid_type!(
             "{:?} is larger than {:?}",
-            dur,
+            duration,
             MAX_TIME_IN_SECS
         ));
     }
     Ok(())
 }
 
-fn tm_to_secs(t: Tm) -> u64 {
-    t.tm_hour as u64 * SECS_PER_HOUR + t.tm_min as u64 * SECS_PER_MINUTE + t.tm_sec as u64
+#[inline]
+fn check_hour(hour: u64) -> Result<u64> {
+    if hour > 838 {
+        Err(invalid_type!("invalid hour: {}, larger than {}", hour, 838))
+    } else {
+        Ok(hour)
+    }
+}
+#[inline]
+fn check_minute(minute: u64) -> Result<u64> {
+    if minute > 59 {
+        Err(invalid_type!(
+            "invalid minute: {}, larger than {}",
+            minute,
+            60
```

Comment:
```suggestion
            59
```

Either change this to `59` or change line 51 to `839`.